sips/pending/_posts/2016-06-25-trailing-commas.md

```diff
@@ -0,0 +1,117 @@
+---
+layout: sip
+disqus: true
+title: SIP-NN - Trailing Commas
+---
+
+# SIP-NN: Trailing Commas
+
+**By: Dale Wijnand**
+
+## History
+
+| Date          | Version       |
+|---------------|---------------|
+| Jun 25th 2016 | Initial Draft |
+
+## Motivation
+
+When using a comma-separated sequence of elements on multiple lines, such as:
+
+{% highlight scala %}
+Seq(
+  foo,
+  bar,
+  baz
+)
+{% endhighlight %}
+
+It is quite inconvenient to remove or comment out any element because one has to think about the fact that the last element mustn't have a trailing comma:
+
+{% highlight scala %}
+Map(
+  foo,
+  bar //,
+//  baz
+)
+{% endhighlight %}
+
+Secondly, it is quite inconvenient to re-order the sequence, for instance if you wanted `baz` before `bar` you need to micromanage which is followed by a comma and which isn't:
+
+{% highlight scala %}
+val xs = Seq(
+  foo,
+  baz   // This isn't going to work
+  bar,
+)
+{% endhighlight %}
+
+Additionally, this means that there is a lot of noise in diffs, such as:
+
+{% highlight diff %}
+@@ -4,7 +4,8 @@
+ Map(
+   foo,
+   bar,
+-  baz
++  baz,
++  quux
+ )
+{% endhighlight %}
+
+There is an open improvement ticket ([SI-4986][]) where this feature was requested, initially in the context of import selectors but later also for other constructs in the syntax.
+
+Some real-world use-cases where elements of a sequence are typically added, removed or moved are:
+
+* invoking constructors or methods (such as `apply` or `copy`) which present a lot of options defined with default values
+* `settings(...)` arguments or elements of `libraryDependencies`, `scalacOptions` or `javaOptions` sequences in sbt
+
+## Design Decisions
+
+There are a number of different elements of the Scala syntax that are comma separated, but instead of changing them all a subset of the more useful ones was chosen:
+
+* tuples
+* argument and parameter groups, including for implicits, for functions, methods and constructors
+* import selectors
+
+From the spec these are:
+
+* SimpleExpr1, ArgumentExprs via Exprs
+* ParamClause, ParamClauses via Params
+* ClassParamClause, ClassParamClauses via ClassParams
+* ImportSelector
+
+The elements that have are not changed are:
+
+* ValDcl, VarDcl, VarDef via ids
+* Type via FunctionArgTypes
+* SimpleType, TypeArgs via Types
+* Expr, ResultExpr via Bindings
+* SimplePattern via Patterns
+* TypeParamClause, FunTypeParamClause
+* ImportExp
+* PatDef
+
+## Implementation
+
+The implementation is a simple change to the parser, allowing for a trailing comma, for the groups detailed above, and has been proposed in [scala/scala#5245][].
+
+## Drawbacks
```

Comment:
Yes, I meant fastparse/scalaparse. My only comment was to include this as a drawback, since updating those tools will require some work.